File: InversePairs/InversePairs.py
class Solution:
    def reversePairs(self, nums: List[int]) -> int:
        if nums ==[]:
            return 0
        temp = [i for i in nums]
        count = 0
        n = len(nums)
        count = self.helper(nums, temp, 0, n-1)
        return count
    def helper(self, nums, temp, start, end):
        if start == end:
            temp[start] = nums[start]
            return 0
        mid = start + (end-start)//2
        left = self.helper(temp, nums, start, mid)
        right = self.helper(temp, nums, mid+1, end)
        
        i = mid
        j = end
        index = end
        count = 0
        while i>=start and j>=mid+1:
            if nums[i] > nums[j]:
                temp[index] = nums[i]
                index-=1
                i-=1
                count+=j-mid
            else:
                temp[index] = nums[j]
                index-=1
                j-=1
        while i>=start:
            temp[index] = nums[i]
            index-=1
            i-=1
        while j>=mid+1:
            temp[index] = nums[j]
            index-=1
            j-=1
        return left + right + count

File: InversePairs/InversePairs.cpp
class Solution {
public:
    int reversePairs(vector<int>& nums) {
        int count = 0;
        if (nums.empty()) return count;
        int n = nums.size();
        vector<int> temp;
        for (int i=0;i<n;i++)
        {
            temp.push_back(nums[i]);
        }
        count = helper(nums, temp, 0, n-1);
        return count;
    }
    int helper(vector<int> &nums, vector<int> &temp, int start, int end)
    {
        if (start==end)
        {
            temp[start] = nums[start];
            return 0;
        }
        int mid = start + (end-start)/2;
        int left = helper(temp, nums, start, mid);
        int right = helper(temp, nums, mid+1, end);

        int i = mid;
        int j = end;
        int index = end;
        int count = 0;
        while (i>=start && j>=mid+1)
        {
            if (nums[i]>nums[j])
            {
                temp[index--] = nums[i--];
                count += j-mid;
            }
            else
            {
                temp[index--] = nums[j--];
            }
        }
        while(i>=start) temp[index--] = nums[i--];
        while(j>=mid+1) temp[index--] = nums[j--];
        return left + right + count;
    }
};
